### ISSUE TITLE
Handling void by nullish coalescing operator

### ISSUE THREAD

**Issue by michaldudak** [microsoft/TypeScript#40613]
**TypeScript Version:**  4.0.2

**Search Terms:** nullish coalescing, void

**Code**

```ts
type PossiblyVoid = () => void | boolean;
let empty : PossiblyVoid = () => { }
let coalesced : boolean = empty() ?? true;
```

**Expected behavior:**
Should not throw a compilation error

**Actual behavior:**
Nullish-coalescing operator does not treat `void` variables as undefined, even though in JS a function that does not return anything is the same as the one returning `undefined`.

I am not entirely sure this is a bug, but intuitively it *feels* like is should behave in a different way. If not, could you please shed some light on the reasons behind it?

**Playground Link:**
https://www.typescriptlang.org/play?ts=4.0.2#code/C4TwDgpgBACg9gZwQSwEYBsQDU7ICZQC8UAFAJREB8UAbrgQD5SpxzoQCGAdgNwBQ7YFAgBbMKCgAuWIhQZs9IqQqFqAbygBfARCEBjOB3YI9EAtJZtOXJaPEhyUAPxOowAE4BXCPyA

**Comment by j-oliveras** [comment]
This is working as intended (see #26234). Imagine:
```ts
ffunction func(): any {
    const d = Math.random();
    if (d < 0.25)
        return 1;
    if (d < 0.5)
        return "-";
    if (d < 0.75)
        return { a: 1};
    return 0;
}

type PossiblyVoid = () => void | boolean;
let empty : PossiblyVoid = func;
let coalesced : boolean = empty() ?? true;
```
[Playground](https://www.typescriptlang.org/play?ts=4.0.2&ssl=1&ssc=18&pln=1&pc=21#code/GYVwdgxgLglg9mABKSAKAlALkQQzAT0QG8AoRcxCBAZykQBNEBeRAWRygAsA6AJz3pwAthgDcZCjGCJUjADyIADNwBMAVnQSKFXgFMoIXkgCM47VJnyl3DVu2I9Bo4gBEAWhdnJ02YgXKAdlt7HX1DJCJcbGMAXy9yR3ClcRiSEih8AAddRAAFOGpqGAAjABt8ADU4GEYWDGYAPkQAN2rGAB9EYrg4Ut08cT66XSFMjMRsfMKS8qqa5mRwCEH9SjgcPuoIXUZsbt7+pBYRsfx6gH5zxCheEF1xIA).

The declared `func` is assignable to `PossiblyVoid` type but the real value of `coalesced` will be `1 | "-" | { a: 1} | 0` instead of a `boolean`.

**Comment by MartinJohns** [comment]
Expanding on what @j-oliveras said:

A function returning `void` doesn't mean it returns nothing, but it returns something that should not be used for anything.

If you intend to have a "nothing" value you should use `undefined`.

**Comment by michaldudak** [comment]
All right, I get it now.

The problem I have with `() => undefined` is that a function must have an explicit `return` statement, so a function like `function fn() : undefned {}` does not compile (even though JS adds an automatic `return undefined` to it.

**Comment by michaldudak** [comment]
Actually, I found an issue that describes this: #36288

**Comment by barryam3** [comment]
> A function returning void doesn't mean it returns nothing, but it returns something that should not be used for anything.

In that case, how come the ternary and or operators work?

```ts
function func(): any {
    const d = Math.random();
    if (d < 0.25)
        return 1;
    if (d < 0.5)
        return "-";
    if (d < 0.75)
        return { a: 1};
    return 0;
}

type PossiblyVoid = () => void | boolean;
let empty : PossiblyVoid = func;
const emptyReturn = empty();
let nullishCoalesced : boolean = emptyReturn ?? true;  // error
let orCoalesced : boolean = emptyReturn || true;  // ok
let ternaryCoalesced: boolean = emptyReturn ? emptyReturn : true;  // ok
```

[Playground](https://www.typescriptlang.org/play?ts=4.0.2&ssl=17&ssc=66&pln=15&pc=1#code/GYVwdgxgLglg9mABKSAKAlALkQQzAT0QG8AoRcxCBAZykQBNEBeRAWRygAsA6AJz3pwAthgDcZCjGCJUjADyIADNwBMAVnQSKFXgFMoIXkgCM47VJnyl3DVu2I9Bo4gBEAWhdnJ02YgXKAdlt7HX1DJCJcbGMAXy9yR3ClcRiSEih8AAddRAAFOGpqGAAjABt8ADU4GEYWDGYAPkQAN2rGAB9EYrg4Ut08cT66XSFMjMRsfMKS8qqa5mRwCHEqMFpEEbH8ACUw5xZNjLESIcQwEFLSmGpOAGE4HD7qCF1GbG7e-qQD0YzdpyQAH5AYgoLwQLpBvpEHBePdHrpnq8Jl0en08AtDjs9kh2p0wRCoXQoLojDhePh4U8XvR3mivpjftiAYgQVj-klsATISQgA)

**Comment by jdforsythe** [comment]
> Expanding on what @j-oliveras said:
> 
> A function returning `void` doesn't mean it returns nothing, but it returns something that should not be used for anything.
> 
> If you intend to have a "nothing" value you should use `undefined`.

Theoretically true, but actually writing a void-returning function that has any value other than `undefined` is tricky.

```ts
function testOne(): void {
  return 1; // type 'number' is not assignable to type 'void'
}

function testTwo(): void {
  console.log('void');
} // returns 'undefined'

function testThree(): void {
  return undefined;
} // returns 'undefined'

function testFour(): void {
  return;
} // returns 'undefined'

function testFive(): boolean | void {
  return;
}

function testSix(): boolean | void {
  return true;
}

function testSeven(): boolean | void {
  return 1; // type '1' is not assignable to type 'boolean | void'
}
```

The only way I can figure out to make it happen is:

```ts
function test(cb: () => any): void {
  return cb();
}

function callback(): boolean {
  return true;
}

test(callback);
```

But since we disallow `any`, it can't actually happen for us. The example works with `any` but not with `unknown` or `never`.

If you disallow `any`, is there any scenario where `void != undefined`?

**Comment by MartinJohns** [comment]
@jdforsythe 

> The only way I can figure out to make it happen is: [...] This just kind of disables the type check, and since we disallow `any`, it can't actually happen for us. The example works with `any` but not with `unknown` or `never`.

The example works with `void`: `function test(cb: () => void): void { return cb(); }`

> We use an explicit `| void` for certain functions to force developers to handle the `undefined` scenario.

Frankly, then you're abusing `void`. If you want to force the developers to handle the `undefined` scenario, then use `undefined`. `void` does **not** mean `undefined`. In a way `void` is the same as `any`, at least when it comes to possible values to deal with.

**Comment by jdforsythe** [comment]
@MartinJohns

Indeed, we are abusing `void`, however using type `undefined` does not provide any benefit.

```ts
function isPositive(input: number): true | undefined {
  if (input < 1) {
    return undefined;
  }

  return true;
}

// val is type `true`
const val = isPositive(-1);
```

Adding the `| undefined` doesn't change anything about the types. There's no warning that `val` might be `undefined`, and in essence any type `T` is `T | undefined`

```ts
const bool: true = undefined;
```

---

There's some inconsistency in the behavior for void. It seems TypeScript only has a problem testing for truthiness if the type is simply `void` but not if it's unioned with another type. In fact, it can be completely wrong.

```ts
const maybeVoid = (): void | number => Math.random() < 0.5 ? alwaysVoid(callback) : 0;

const alwaysVoid = (cb: () => void): void => cb();
const callback = (): boolean => true;

function try1() {
  const rec = alwaysVoid(callback);

  if (!rec) { // An expression of type 'void' cannot be tested for truthiness
    return;
  }
}

function try2() {
  const rec: number | void = maybeVoid();

  if (!rec) {
    console.log('number | void, but actually can only be number here', rec);

    return;
  }

  console.log('number, but actually can only be void here', rec);
}
```

In `try2` TypeScript has no problem checking truthiness of `!rec` even though it could be `void`, and in this case gets it exactly backwards. Inside the `if` block, it types `rec` as `number | void` even though it could only be a number (0) there, and after the `if` it types `rec` as `number` even though it can only be `void` (`true`).

**Comment by MartinJohns** [comment]
> Adding the | undefined doesn't change anything about the types. There's no warning that val might be undefined, and in essence any type T is T | undefined

It makes a difference if you enable `strictNullChecks`, a feature existing since TypeScript 2.0 (released in 2016).
